java/main/jglome/Glome.java

```diff
@@ -0,0 +1,363 @@
+   * @param cnt number of previously sent messages from the peer to the user.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [0..255] range.
+   * @throws WrongTagException if the peer's tag has invalid length (less than {@code
+   * minPeerTagLength} or more than {@code MAX_TAG_LENGTH}) or it's not equal to the prefix of a
+   * correct tag.
+   */
+  public void checkTag(byte[] peerTag, byte[] msg, int cnt)
+      throws CounterOutOfBoundsException, WrongTagException {
+    if (peerTag.length < minPeerTagLength || peerTag.length > MAX_TAG_LENGTH) {
+      throw new WrongTagException(
+          "The received tag has invalid length. Expected " + minPeerTagLength + " or more, got "
+              + peerTag.length + ".");
+    }
+
+    byte[] truncatedTag = Arrays.copyOf(generateTag(msg, cnt, this.peerMacKey), peerTag.length);
+    if (!Arrays.equals(peerTag, truncatedTag)) {
+      throw new WrongTagException("The received tag doesn't match the expected tag.");
+    }
+  }
+
+  /**
+   * Common part of the constructors. Initialises peer's public key, minimum length of a peer's tag,
+   * base point for Curve25519 curve. Calculates user's and peer's MAC keys.
+   *
+   * @param peerKey peer's public key.
+   * @param minPeerTagLength minimum length of a peer's tag.
+   * @throws MinPeerTagLengthOutOfBoundsException if {@code minPeerTagLength} is out of [1..32]
+   * range.
+   */
+  private void initMacKeys(XECPublicKey peerKey, int minPeerTagLength)
+      throws MinPeerTagLengthOutOfBoundsException {
+    this.peerKey = peerKey;
+    if (minPeerTagLength < MIN_TAG_LENGTH || minPeerTagLength > MAX_TAG_LENGTH) {
+      throw new MinPeerTagLengthOutOfBoundsException(
+          "minPeerTagLength argument should be in [1..32] range. Got " + minPeerTagLength + ".");
+    }
+    this.minPeerTagLength = minPeerTagLength;
+
+    KeyAgreement ka;
+    try {
+      ka = KeyAgreement.getInstance("XDH");
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError(e.getMessage());
+    }
+
+    try {
+      ka.init(userPrivateKey());
+      ka.doPhase(peerKey, true);
+    } catch (InvalidKeyException e) {
+      throw new AssertionError(e.getMessage());
+    }
+
+    byte[] sharedSecret = ka.generateSecret();
+    byte[] userMac = reverseByteArray(
+        getU32Bytes(userPublicKey())); // in order to have <i>little-endian</i> byte-order;
+    byte[] peerMac = reverseByteArray(getU32Bytes(peerKey)); // the same
+
+    this.userMacKey = getMacKey(sharedSecret, peerMac, userMac);
+    this.peerMacKey = getMacKey(sharedSecret, userMac, peerMac);
+  }
+
+  private byte[] getU32Bytes(XECPublicKey pk) {
+    byte[] userUInBytes = pk.getU().toByteArray();
+    byte[] userU32Bytes = new byte[32];
+    System.arraycopy(userUInBytes, 0, userU32Bytes, 32 - userUInBytes.length, userUInBytes.length);
+
+    return userU32Bytes;
+  }
+
+  private XECPublicKey derivePublicKeyFromPrivate(XECPrivateKey privateKey) {
```

Comment:
I think the correct way to do this is over [java.security.KeyFactory](https://docs.oracle.com/javase/7/docs/api/java/security/KeyFactory.html).